Flux panics when .gitconfig has multiline configs

I have the following .gitconfig:

```
[alias]
        l = log --pretty=oneline -n 20 --graph --abbrev-commit

        mpr = "!f() { \
          if [ $(printf \"%s\" \"$1\" | grep '^[0-9]\\+$' > /dev/null; printf $?) -eq 0 ]; then \
            git fetch origin refs/pull/$1/head:pr/$1 && \
            git rebase master pr/$1 && \
            git checkout master && \
            git merge pr/$1 && \
            git branch -D pr/$1 && \
            git commit --amend -m \"$(git log -1 --pretty=%B)\n\nCloses #$1.\"; \
          fi \
        }; f"

[color]
        ui = auto
```

And Flux panics with:

```
 % fluxctl identity
panic: runtime error: index out of range

goroutine 1 [running]:
main.userGitconfigMap(0xc4202e6300, 0x2e8, 0x2e8)
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/args.go:60 +0x232
main.getUserGitconfig(0x3d)
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/args.go:48 +0x175
main.AddCauseFlags(0xc4202ee6c0, 0xc420096a18)
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/args.go:15 +0x26
main.(*controllerReleaseOpts).Command(0xc4200969a0, 0xc4200969a0)
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/release_cmd.go:46 +0x1a5
main.(*rootOpts).Command(0xc4202d9440, 0xc4202d9440)
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/root_cmd.go:62 +0x53b
main.run(0xc42000c1d0, 0x1, 0x1, 0xc4200001a0)
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/main.go:12 +0x4d
main.main()
        /home/deploy/go/src/github.com/weaveworks/flux/cmd/fluxctl/main.go:29 +0x64
```

 **cmd/fluxctl/args.go**  is pulling the entire .gitconfig file, when it only needs to parse user.name & user.email fields.

--- a/cmd/fluxctl/args.go
+++ b/cmd/fluxctl/args.go
@@ -3,8 +3,9 @@ package main
 import (
 	"bytes"
 	"fmt"
+        "io/ioutil"
 	"os/exec"
-	"strings"
+        "strings"
 
 	"github.com/spf13/cobra"
 
@@ -12,16 +13,15 @@ import (
 )
 
 func AddCauseFlags(cmd *cobra.Command, opts *update.Cause) {
-	authorInfo := getUserGitconfig()
-	username := getCommitAuthor(authorInfo)
+	username := getCommitAuthor()
 
 	cmd.Flags().StringVarP(&opts.Message, "message", "m", "", "attach a message to the update")
 	cmd.Flags().StringVar(&opts.User, "user", username, "override the user reported as initiating the update")
 }
 
-func getCommitAuthor(authorInfo map[string]string) string {
-	userName := authorInfo["user.name"]
-	userEmail := authorInfo["user.email"]
+func getCommitAuthor() string {
+	userName := getUserGitConfigValue("user.name")
+	userEmail := getUserGitConfigValue("user.email")
 
 	switch {
 	case userName != "" && userEmail != "":
@@ -34,39 +34,17 @@ func getCommitAuthor(authorInfo map[string]string) string {
 	return ""
 }
 
-func getUserGitconfig() map[string]string {
+var execCommand = exec.Command
+func getUserGitConfigValue(arg string) string {
 	var out bytes.Buffer
-	userGitconfig := make(map[string]string)
-	cmd := exec.Command("git", "config", "--list")
+	cmd := execCommand("git", "config", "--get", "--null", arg)
 	cmd.Stdout = &out
+	cmd.Stderr = ioutil.Discard
 
 	err := cmd.Run()
 	if err != nil {
-		return userGitconfig
+		return ""
 	}
 	res := out.String()
-	return userGitconfigMap(res)
-}
-
-func userGitconfigMap(s string) map[string]string {
-	c := make(map[string]string)
-	lines := splitList(s)
-	for _, l := range lines {
-		if l == "" {
-			continue
-		}
-		prop := strings.SplitN(l, "=", 2)
-		p := strings.TrimSpace(prop[0])
-		v := strings.TrimSpace(prop[1])
-		c[p] = v
-	}
-	return c
-}
-
-func splitList(s string) []string {
-	outStr := strings.TrimSpace(s)
-	if outStr == "" {
-		return []string{}
-	}
-	return strings.Split(outStr, "\n")
+	return strings.Trim(res, "\x00")
 }

--- a/cmd/fluxctl/args_test.go
+++ b/cmd/fluxctl/args_test.go
@@ -1,122 +1,62 @@
 package main
 
 import (
-	"reflect"
+        "fmt"
 	"testing"
+	"os"
+	"os/exec"
 )
 
-var gitConfig = `push.default=simple
-    merge.conflictstyle=diff3
-    pull.ff=only
-    core.repositoryformatversion=0
-    core.filemode=true
-    core.bare=false
-    `
-
-func gitConfigMap(input map[string]string) map[string]string {
-	res := map[string]string{
-		"push.default":                 "simple",
-		"merge.conflictstyle":          "diff3",
-		"pull.ff":                      "only",
-		"core.repositoryformatversion": "0",
-		"core.filemode":                "true",
-		"core.bare":                    "false",
-	}
-	for k, v := range input {
-		res[k] = v
-	}
-	return res
+func helperCommand(command string, s ...string) (cmd *exec.Cmd) {
+	cs := []string{"-test.run=TestHelperProcess", "--", command}
+	cs = append(cs, s...)
+	cmd = exec.Command(os.Args[0], cs...)
+	cmd.Env = []string{"GO_WANT_HELPER_PROCESS=1"}
+	return cmd
 }
 
-func checkUserGitconfigMap(t *testing.T, d string, expected map[string]string) {
-	userGitconfigInfo := userGitconfigMap(d)
+func TestHelperProcess(t *testing.T) {
+	if os.Getenv("GO_WANT_HELPER_PROCESS") != "1" {
+		return
+	}
+	defer os.Exit(0)
 
-	if len(userGitconfigInfo) != len(expected) {
-		t.Fatalf("got map with %d keys instead of expected %d",
-			len(userGitconfigInfo), len(expected))
+
+	args := os.Args
+	for len(args) > 0 {
+		if args[0] == "--" {
+			args = args[1:]
+			break
+		}
+		args = args[1:]
 	}
-	if !reflect.DeepEqual(userGitconfigInfo, expected) {
-		t.Fatal("result map does not match expected structure")
+	if len(args) == 0 {
+		t.Fatalf("No command\n")
 	}
+
+	_, args = args[0], args[1:]
+        for _, a := range args {
+            if a == "user.name" {
+    		fmt.Fprintf(os.Stdout, "Jane Doe")
+            } else if a == "user.email" {
+    		fmt.Fprintf(os.Stdout, "jd@j.d")
+            }
+        }
 }
 
-func checkAuthor(t *testing.T, input map[string]string, expected string) {
-	author := getCommitAuthor(input)
+func checkAuthor(t *testing.T, input string, expected string) {
+    	execCommand = helperCommand
+    	defer func(){ execCommand = exec.Command }()
+	author := getUserGitConfigValue(input)
 	if author != expected {
 		t.Fatalf("author %q does not match expected value %q", author, expected)
 	}
 }
 
-func TestUserGitconfigMap_EmptyString(t *testing.T) {
-	d := ""
-	expected := map[string]string{}
-	checkUserGitconfigMap(t, d, expected)
-}
-
-func TestUserGitconfigMap(t *testing.T) {
-	d := gitConfig
-	expected := gitConfigMap(nil)
-	checkUserGitconfigMap(t, d, expected)
-}
-
-func TestUserGitconfigMap_WithEmptyLines(t *testing.T) {
-	d := `
-    user.name=Jane Doe
-
-    ` + gitConfig + `
-    `
-	expected := gitConfigMap(map[string]string{
-		"user.name": "Jane Doe",
-	})
-	checkUserGitconfigMap(t, d, expected)
-}
-
-func TestUserGitconfigMap_WithNoKeys(t *testing.T) {
-	d := `
-	`
-	expected := map[string]string{}
-	checkUserGitconfigMap(t, d, expected)
-}
-
-func TestGetCommitAuthor_BothNameAndEmail(t *testing.T) {
-	input := gitConfigMap(map[string]string{
-		"user.name":  "Jane Doe",
-		"user.email": "jd@j.d",
-	})
-	checkAuthor(t, input, "Jane Doe <jd@j.d>")
-}
-
 func TestGetCommitAuthor_OnlyName(t *testing.T) {
-	input := gitConfigMap(map[string]string{
-		"user.name": "Jane Doe",
-	})
-	checkAuthor(t, input, "Jane Doe")
+	checkAuthor(t, "user.name", "Jane Doe")
 }
 
 func TestGetCommitAuthor_OnlyEmail(t *testing.T) {
-	input := gitConfigMap(map[string]string{
-		"user.email": "jd@j.d",
-	})
-	checkAuthor(t, input, "jd@j.d")
-}
-
-func TestGetCommitAuthor_NoNameNoEmail(t *testing.T) {
-	input := gitConfigMap(nil)
-	checkAuthor(t, input, "")
-}
-
-func TestGetCommitAuthor_NameAndEmptyEmail(t *testing.T) {
-	input := gitConfigMap(map[string]string{
-		"user.name":  "Jane Doe",
-		"user.email": "",
-	})
-	checkAuthor(t, input, "Jane Doe")
-}
-
-func TestGetCommitAuthor_EmailAndEmptyName(t *testing.T) {
-	input := gitConfigMap(map[string]string{
-		"user.name":  "",
-		"user.email": "jd@j.d",
-	})
-	checkAuthor(t, input, "jd@j.d")
+	checkAuthor(t, "user.email", "jd@j.d")
 }

